python/pyarrow/_flight.pyx

```diff
@@ -474,6 +485,148 @@ cdef class GeneratorStream(FlightDataStream):
+                             "ServerAuthHandler.authenticate")
+        check_status(self.reader.Read(token))
+        return token
+
+    cdef void poison(self):
+        """Prevent further usage of this object.
+
+        This object is constructed by taking a pointer to a reference,
+        so we want to make sure Python users do not access this after
+        the reference goes away.
+        """
+        self.reader = NULL
+
+    @staticmethod
+    cdef ServerAuthReader wrap(CServerAuthReader& reader):
+        cdef ServerAuthReader result = \
+            ServerAuthReader.__new__(ServerAuthReader)
+        result.reader = &reader
+        return result
+
+
+cdef class ServerAuthSender:
+    """A writer for messages to the client during an auth handshake."""
+    cdef:
+        CServerAuthSender* sender
+
+    cdef write(self, token):
+        cdef c_string c_token = token
+        # TODO: worth acquiring lock?
+        if not self.sender:
+            raise ValueError("Cannot use ServerAuthSender outside "
+                             "ServerAuthHandler.authenticate")
+        check_status(self.sender.Write(token))
+        return token
+
+    cdef void poison(self):
+        """Prevent further usage of this object.
+
+        This object is constructed by taking a pointer to a reference,
+        so we want to make sure Python users do not access this after
+        the reference goes away.
+        """
+        self.sender = NULL
+
+    @staticmethod
+    cdef ServerAuthSender wrap(CServerAuthSender& sender):
+        cdef ServerAuthSender result = \
+            ServerAuthSender.__new__(ServerAuthSender)
+        result.sender = &sender
+        return result
+
+
+cdef class ClientAuthReader:
+    """A reader for messages from the server during an auth handshake."""
+    cdef:
+        CClientAuthReader* reader
+
+    cdef read(self):
+        cdef c_string token
+        if not self.reader:
+            raise ValueError("Cannot use ClientAuthReader outside "
+                             "ClientAuthHandler.authenticate")
+        check_status(self.reader.Read(token))
+        return token
+
+    cdef void poison(self):
+        """Prevent further usage of this object.
+
+        This object is constructed by taking a pointer to a reference,
+        so we want to make sure Python users do not access this after
+        the reference goes away.
+        """
+        self.reader = NULL
+
+    @staticmethod
+    cdef ClientAuthReader wrap(CClientAuthReader& reader):
+        cdef ClientAuthReader result = \
+            ClientAuthReader.__new__(ClientAuthReader)
+        result.reader = &reader
+        return result
+
+
+cdef class ClientAuthSender:
+    """A writer for messages to the server during an auth handshake."""
+    cdef:
+        CClientAuthSender* sender
+
+    cdef write(self, token):
+        cdef c_string c_token = token
+        # TODO: worth acquiring lock?
```

Comment:
Which lock is this talking about?

Reply:
Ah, I'm debating whether it's worth putting a lock on the Python side to avoid a race between write() and poison().

Reply:
I don't think so, no.